saja/task_manager.py

```diff
@@ -0,0 +1,199 @@
+from utils import get_valid_input
+
+class TaskManager:
+    _instance = None
+
+    """Class to manage tasks including adding, viewing, updating, and deleting."""
+
+    def __new__(cls, *args, **kwargs):
+        """
+        Ensure only one instance of TaskManager is created.
+        """
+        if cls._instance is None:
+            cls._instance = super(TaskManager, cls).__new__(cls)
+        return cls._instance
+
+    def __init__(self, data_file='data.json'):
+        """
+        Initialize the TaskManager with a specified data file.
+        
+        Parameters:
+        - data_file: The path to the JSON file storing task data.
+        """
+        if not hasattr(self, 'initialized'):  # Prevent reinitialization
+            self.data_file = data_file
+            self.tasks = []
+            self.urgent_tasks = []
+            self.load_tasks()
+            self.initialized = True  # Mark as initialized
+
+    def load_tasks(self):
+        """Load tasks from the JSON data file into the manager."""
+        try:
+            with open(self.data_file, 'r') as f:
+                data = json.load(f)
+                for item in data:
+                    status = Status[item['status']]
+                    if 'priority' in item:
+                        priority = Priority[item['priority'].capitalize()]
+                        task = UrgentTask(
+                            title=item['title'],
+                            description=item['description'],
+                            due_date=item['due_date'],
+                            status=status,
+                            priority=priority
+                        )
+                        self.urgent_tasks.append(task)
+                    else:
+                        task = Task(
+                            title=item['title'],
+                            description=item['description'],
+                            due_date=item['due_date'],
+                            status=status
+                        )
+                        self.tasks.append(task)
+        except FileNotFoundError:
+            print(f"File {self.data_file} not found. Starting with an empty task list.")
+        except json.JSONDecodeError:
+            print("Error decoding JSON data. Starting with an empty task list.")
+
+    def save_tasks(self):
+        """Save the current tasks back to the JSON data file."""
+        data = []
+        for task in self.tasks + self.urgent_tasks:
+            task_data = {
+                'title': task.title,
+                'description': task.description,
+                'due_date': task.due_date.strftime('%Y-%m-%d'),
+                'status': task.status.name
+            }
+            if isinstance(task, UrgentTask):
+                task_data['priority'] = task.priority.name
+            data.append(task_data)
+
+        with open(self.data_file, 'w') as f:
+            json.dump(data, f, indent=4)
+
+    def add_task(self):
```

Comment:
maybe we need to put this code under try/catch block ?
because the validation can raise an exceptions.